### PR TITLE
[WIP] docs: Update protodoc

### PR DESCRIPTION
**Pull request by phlax**
- use templates
- optimizations
- cleanups

Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: tools/api_proto_plugin/plugin.py [hunk 1]
```diff
@@ -88,9 +86,9 @@ def plugin(output_descriptors):
...
             ps.print_stats()
             stats_file.content = stats_stream.getvalue()
-        # Also include the original FileDescriptorProto as text proto, this is
-        # useful when debugging.
-        descriptor_file = response.file.add()
-        descriptor_file.name = file_proto.name + ".descriptor.proto"
-        descriptor_file.content = str(file_proto)
+            # Also include the original FileDescriptorProto as text proto, this is
+            # useful when debugging.
+            descriptor_file = response.file.add()
```
Comment (htuch) [inline tools/api_proto_plugin/plugin.py:91]: This has been moved underneath the `cprofile_enabled`?

**Review by htuch — COMMENTED** [review]
@phlax I agree with all these changes, but can you please make the review a bit more tractable and separate out different parts, e.g. the templating, the general protodoc cleanups and so on. It's very hard to do a before/after comparison when there are many moving pieces. Thanks. Also, verifying that docs are the same in some automatic way across the changes would increase confidence in correctness.

File: tools/api_proto_plugin/plugin.py (hunk 1 above)
Comment (phlax) [reply tools/api_proto_plugin/plugin.py:91]: yep - for testing purposes

not sure what is best here - i think in the main we dont want the io of creating this when its not used (iiuc it mostly isnt)

perhaps we can add another os.environ flag - `DUMP_PROTO_DESCRIPTORS` or similar

**Comment by phlax** [comment]
> please make the review a bit more tractable and separate out different parts

just to flag this is my currently-not-working-as-we-might-hope r&d/dev branch

essentially i got this far and mostly got it working and set out to do exactly that (ie separate out different parts)

the first part was altering the schema (#21528 )

next up is adding a macro for creating a (compiled) jinja env (#21533 )

this part also has some upstream components - most of which are landed https://github.com/envoyproxy/pytooling/pull/549 (just iterating some cleanups there)

once this is out of the way the other parts are (in a yet to be figured out order):

- add an upstream `TraversalAdapter` and the hack to allow it to be used
- filter envoy/xds protos earlier in api_proto_plugin
- minor cleanups for protodoc.py

once all of those are out of the way this PR should only have the main changes to `protodoc.py` - the invasive ones - and at that point we can look at some diff/regression testing to make sure nothing breaks

**Comment by htuch** [comment]
Sounds like a great plan, thanks.

**Comment by phlax** [comment]
as an update to this i have finally got the templates ~working and correct - need to clean them up a bit yet

i will start splitting out some of the protodoc and related changes this week

File: tools/api_proto_plugin/plugin.py [hunk 2]
```diff
@@ -82,9 +85,9 @@ def plugin(output_descriptors):
             stats_file.name = file_proto.name + '.profile'
             ps.print_stats()
             stats_file.content = stats_stream.getvalue()
-        # Also include the original FileDescriptorProto as text proto, this is
-        # useful when debugging.
-        descriptor_file = response.file.add()
-        descriptor_file.name = file_proto.name + ".descriptor.proto"
-        descriptor_file.content = str(file_proto)
+            # Also include the original FileDescriptorProto as text proto, this is
```
Comment (phlax) [inline tools/api_proto_plugin/plugin.py:88]: dont do this here

File: tools/api_proto_plugin/type_context.py [hunk 3]
```diff
@@ -89,7 +87,8 @@ def leading_comment_path_lookup(self, path):
         """
         location = self.location_path_lookup(path)
         if location is not None:
-            return Comment(location.leading_comments, self.file_level_annotations)
+            comments = "\n".join(s[1:] for s in location.leading_comments.split("\n"))
```
Comment (phlax) [inline tools/api_proto_plugin/type_context.py:90]: can this be done in the template

File: tools/protodoc/BUILD [hunk 4]
```diff
@@ -111,6 +116,7 @@ py_binary(
         "@com_github_cncf_udpa//xds/annotations/v3:pkg_py_proto",
         "@com_google_protobuf//:protobuf_python",
         requirement("envoy.code.check"),
+        dev_requirement("remote_pdb"),
```
Comment (phlax) [inline tools/protodoc/BUILD:119]: dev

File: tools/protodoc/generate_empty.py [hunk 5]
```diff
@@ -7,25 +7,25 @@
...
     path = pathlib.Path(api_extensions_root, extension_root, 'empty', extension_root.name + '.rst')
     path.parent.mkdir(parents=True, exist_ok=True)
     description = details.get('description', '')
     reflink = ''
     if 'ref' in details:
         reflink = '%s %s.' % (
-            details['title'], protodoc.format_internal_link(
-                'configuration overview', details['ref']))
+            details['title'],
+            _protodoc.format_internal_link('configuration overview', details['ref']))
```
Comment (phlax) [inline tools/protodoc/generate_empty.py:23]: filter ?

File: tools/protodoc/protodoc.py [hunk 6]
```diff
@@ -55,23 +47,57 @@
...
+    field_proto.TYPE_STRING: 'string',
+    field_proto.TYPE_BYTES: 'bytes'
+}
+
+PRETTY_LABEL_NAMES = {field_proto.LABEL_OPTIONAL: '', field_proto.LABEL_REPEATED: '**repeated** '}
+
+import contextlib
+
+
+@contextlib.contextmanager
```
Comment (phlax) [inline tools/protodoc/protodoc.py:87]: remove this

File: tools/protodoc/protodoc.py [hunk 7]
```diff
@@ -414,289 +194,488 @@ def field_cross_ref_label(field_name):
...
+            rpc, 'https://cloud.google.com/natural-language/docs/reference/rpc/google.rpc#%s'
+            % rpc.lower())
 
-        message = anchor + header + proto_link + formatted_leading_comment + format_message_as_json(
-            type_context, msg_proto) + format_message_as_definition_list(
-                type_context, msg_proto,
-                self.protodoc_manifest) + '\n'.join(nested_msgs) + '\n' + '\n'.join(nested_enums)
-        error = self.backticks_check(message)
-        if error:
-            logger.warning(f"Bad RST ({msg_proto.name}): {error}")
```
Comment (phlax) [inline tools/protodoc/protodoc.py:656]: this still needs to happen somewhere

**Pull request closed without merging**